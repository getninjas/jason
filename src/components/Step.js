import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Field from './Field';
import Button from './Button';
import Factory from './Factory';
import { display, addHeaderMarkup } from '../helpers/step';

const propTypes = {
  onSubmit: PropTypes.func.isRequired,
  onFieldChange: PropTypes.func.isRequired,
  zipcodeUrlService: PropTypes.string.isRequired,
  formName: PropTypes.string.isRequired,
  buttonText: PropTypes.string.isRequired,
  fields: PropTypes.array.isRequired,
  headerMarkup: PropTypes.string,
  visible: PropTypes.bool,
  isLast: PropTypes.bool,
};

const defaultProps = {
  visible: false,
  isLast: false,
};

export default class Step extends Component {
  render() {
    const { buttonText, headerMarkup, fields } = this.props;

    return (
      <fieldset className="form__container inputs" style={{ display: display(this.props.visible) }}>
        { addHeaderMarkup(headerMarkup) }

        {
          fields.map((item, index) =>
            <Field
              key={`field-${index}`}
              label={item.title}
              id={item.id}
              errorMessage={item.errorMessage}
              wrapperStyle={item.wrapperStyle} >
              {
                Factory.getComponent({
                  item,
                  index,
                  onFieldChange: this.props.onFieldChange,
                  formName: this.props.formName,
                  zipcodeUrlService: this.props.zipcodeUrlService,
                })
              }
            </Field>,
          )
        }

        <Button isSubmit={this.props.isLast} handleButtonClick={this.props.onSubmit}>
          {buttonText}
        </Button>
      </fieldset>
    );
  }
}

Step.propTypes = propTypes;
Step.defaultProps = defaultProps;
