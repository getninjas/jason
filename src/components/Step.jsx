import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Field from './Field';
import Button from './Button';
import Factory from './Factory';

const propTypes = {
  handleButtonClick: PropTypes.func.isRequired,
  onFieldChange: PropTypes.func.isRequired,
  zipcodeUrlService: PropTypes.string.isRequired,
  formName: PropTypes.string.isRequired,
  buttonText: PropTypes.string.isRequired,
  fields: PropTypes.array.isRequired,
  headerMarkup: PropTypes.string,
  visible: PropTypes.bool,
  isLast: PropTypes.bool,
};

const defaultProps = {
  visible: false,
  isLast: false,
};

export default class Step extends Component {
  display({ visible }) {
    return visible ? 'block' : 'none';
  }

  addHeaderMarkup(headerMarkup) {
    return headerMarkup ? <div className="__headerMarkup__" dangerouslySetInnerHTML={this._createMarkup(headerMarkup)} /> : '';
  }

  render() {
    const { buttonText, headerMarkup, fields } = this.props;

    return (
      <fieldset className="form__container inputs" style={{ display: this.display(this.props) }}>
        { this.addHeaderMarkup(headerMarkup) }

        {
          fields.map((item, index) => {
            return (
              <Field
                key={`field-${index}`}
                label={item.title}
                id={item.id}
                errorMessage={item.errorMessage}>
                {
                  Factory.getComponent({
                    item,
                    index,
                    onFieldChange: this.props.onFieldChange,
                    formName: this.props.formName,
                    zipcodeUrlService: this.props.zipcodeUrlService
                  })
                }
              </Field>
            )
          })
        }

        <Button isSubmit={this.props.isLast} handleButtonClick={this.props.handleButtonClick}>{buttonText}</Button>
      </fieldset>
    );
  }

  _createMarkup(html) {
    return { __html: html };
  }
}

Step.propTypes = propTypes;
Step.defaultProps = defaultProps;
