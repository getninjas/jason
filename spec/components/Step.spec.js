import React from 'react';
import renderer from 'react-test-renderer';
import Step from '../../src/components/Step';
import formData from '../../src/form.json';
import { enzymeConfig, shallow } from '../enzymeConfig';

enzymeConfig();

describe('Step', () => {
  it('renders with props', () => {
    const step = formData.form.steps[0];

    const component = renderer.create(
      <Step
        key={'step'}
        buttonText={'next step'}
        fields={step.fields}
        formName={'formNameTest'}
        onSubmit={() => {}}
        isLast={false}
        step={step}
        zipcodeUrlService={formData.form.zipcodeUrlService}
        onFieldChange={() => {}}
        onFieldBlur={() => {}}
        visible={true}
      />,
    );

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('shows header html when exists', () => {
    const step = formData.form.steps[0];

    const component = shallow(
      <Step
        key={'step'}
        buttonText={'next step'}
        fields={step.fields}
        formName={'formNameTest'}
        onSubmit={() => {}}
        headerMarkup={step.headerMarkup}
        isLast={false}
        step={step}
        zipcodeUrlService={formData.form.zipcodeUrlService}
        onFieldChange={() => {}}
        onFieldBlur={() => {}}
        visible={true}
      />,
    );

    const result = component.html();

    expect(result).toContain('widget__title');
  });

  it('shows inputs, select and textarea inner component', () => {
    const step = formData.form.steps[0];

    const component = shallow(
      <Step
        key={'step'}
        buttonText={'next step'}
        fields={step.fields}
        formName={'formNameTest'}
        onSubmit={() => {}}
        isLast={false}
        step={step}
        onFieldChange={() => {}}
        onFieldBlur={() => {}}
        zipcodeUrlService={formData.form.zipcodeUrlService}
        visible={true}
      />,
    );

    const result = component.html();

    expect(result.includes('1_name_select')).toBe(true);
    expect(result.includes('2_name_text')).toBe(true);
    expect(result.includes('3_name_textarea')).toBe(true);
    expect(result.includes('breadcrumb')).toBe(false);
  });

  describe('.addHeaderMarkup', () => {
    it('does not add header markup', () => {
      const step = formData.form.steps[0];

      const component = shallow(
        <Step
          key={'step'}
          buttonText={'next step'}
          fields={step.fields}
          formName={'formNameTest'}
          onSubmit={() => {}}
          isLast={false}
          step={step}
          onFieldChange={() => {}}
          onFieldBlur={() => {}}
          zipcodeUrlService={formData.form.zipcodeUrlService}
          visible={true}
        />,
      );

      const result = component.html();

      expect(result.includes('__headerMarkup__')).toBe(false);
    });

    it('adds header markup', () => {
      const step = formData.form.steps[0];

      const component = shallow(
        <Step
          key={'step'}
          buttonText={'next step'}
          fields={step.fields}
          formName={'formNameTest'}
          onSubmit={() => {}}
          headerMarkup={step.headerMarkup}
          isLast={false}
          step={step}
          onFieldChange={() => {}}
          onFieldBlur={() => {}}
          zipcodeUrlService={formData.form.zipcodeUrlService}
          visible={true}
        />,
      );

      const result = component.html();

      expect(result.includes('__headerMarkup__')).toBe(true);
      expect(result.includes('widget__title')).toBe(true);
    });
  });
});
