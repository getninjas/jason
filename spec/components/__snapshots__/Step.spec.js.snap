// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step renders with props 1`] = `
<fieldset
  className="form__container inputs"
  style={
    Object {
      "display": "block",
    }
  }
>
  
  <div
    className="form__field form__field--fluid input"
  >
    <label
      className="form__label"
      htmlFor="1_id"
    >
      Qual é o tipo de serviço que você precisa?
    </label>
    <select
      className="form__input"
      defaultValue={null}
      id="formNameTest-1_id"
      name="1_name_select"
      onChange={[Function]}
      required="true"
    >
      <option
        value=""
      >
        Placeholder
      </option>
      <option
        value={7117}
      >
        Construção
      </option>
      <option
        value={7118}
      >
        Reformas
      </option>
      <option
        value={7119}
      >
        Instalações
      </option>
      <option
        value={7120}
      >
        Troca
      </option>
      <option
        value={7121}
      >
        Outros
      </option>
    </select>
    <span
      className="form__message form__message--invalid space-element-small error"
    >
      
    </span>
  </div>
  <div
    className="form__field form__field--fluid input"
  >
    <label
      className="form__label"
      htmlFor="2_id"
    >
      Especifique melhor a metragem do serviço
    </label>
    <input
      className="form__input"
      id="formNameTest-2_id"
      maxLength={255}
      minLength={3}
      name="2_name_text"
      onChange={[Function]}
      placeholder="Bora pra action"
      required="true"
      title=""
      type="text"
      value=""
    />
    <span
      className="form__message form__message--invalid space-element-small error"
    >
      
    </span>
  </div>
  <div
    className="form__field form__field--fluid input"
  >
    <label
      className="form__label"
      htmlFor="3_id"
    >
      Informações adicionais
    </label>
    <textarea
      className="form__input"
      id="formNameTest-3_id"
      maxLength={255}
      minLength={3}
      name="3_name_textarea"
      onChange={[Function]}
      placeholder="Descreva o que você precisa"
      required={true}
      title=""
      value=""
    />
    <span
      className="form__message form__message--invalid space-element-small error"
    >
      
    </span>
  </div>
  <div
    className="form__field form__field--fluid input"
  >
    <label
      className="form__label"
      htmlFor="4_id"
    >
      Zipcode
    </label>
    <a
      className="form__label-link"
      href="http://www.buscacep.correios.com.br"
      rel="noopener noreferrer"
      target="_blank"
    >
      Não lembra seu CEP?
    </a>
    <input
      className="form__input form__input--filled"
      id="4_id"
      name="4_zipcode"
      onKeyUp={[Function]}
      placeholder="00000-000"
      required={true}
      type="tel"
    />
    <span
      className="full-address"
    >
      
    </span>
    <input
      id="street"
      name="street"
      type="hidden"
      value=""
    />
    <input
      id="neighborhood"
      name="neighborhood"
      type="hidden"
      value=""
    />
    <input
      id="city"
      name="city"
      type="hidden"
      value=""
    />
    <input
      id="uf"
      name="uf"
      type="hidden"
      value=""
    />
    <span
      className="form__message form__message--invalid space-element-small error"
    >
      Required field
    </span>
  </div>
  <button
    className="btn btn--regular btn--high-contrast btn--fluid space-box-medium"
    onClick={[Function]}
    type="button"
  >
    next step
  </button>
</fieldset>
`;
